server/sources.py

```diff
@@ -0,0 +1,302 @@
+        else:
+            for p1, p2, score in self.iter_tricol(by='object'):
+                yield p1[by], p2[by], score
+
+
+def read_tricol_scores(file_path, net1=None, net2=None, by='name', **kwargs):
+    if 'delimiter' not in kwargs:
+        kwargs['delimiter'] = '\t'
+    return TricolScoreMatrix(util.iter_csv(file_path, **kwargs), net1=net1, net2=net2, by=by)
+
+def identity_scores(net, by='name'):
+    return TricolScoreMatrix([(v, v, 1) for v in net.iter_vertices(by=by)], by=by)
+
+
+class IsobaseLocal(object):
+    def __init__(self, base_path):
+        self.base_path = base_path
+
+    def _check_valid_species(self, species_name):
+        if not species_name.isalnum():
+            raise ValueError(f'invalid species name: {species_name}')
+
+    def get_network(self, species_name):
+        self._check_valid_species(species_name)
+        species_path = path.join(self.base_path, f'{species_name}.tab')
+
+        if not path.isfile(species_path):
+            raise IOError(f'network file for {species_name} was not found')
+
+        return read_net_tsv_edgelist(species_path)
+
+    def get_score_matrix(self, species1_name, species2_name=None, net1=None, net2=None):
+        self._check_valid_species(species1_name)
+
+        if species2_name is None:
+            matrix_path = path.join(self.base_path, f'{species1_name}-blast.tab')
+
+            if not path.isfile(matrix_path):
+                raise IOError(f'score matrix file for {species1_name} was not found')
+        else:
+            self._check_valid_species(species2_name)
+            matrix_path = path.join(self.base_path, f'{species1_name}-{species2_name}-blast.tab')
+
+            if not path.isfile(matrix_path):
+                raise IOError(f'score matrix file for {species1_name}-{species2_name} was not found')
+
+        return read_tricol_scores(matrix_path, net1=net1, net2=net2)
+
+    def get_ontology_mapping(self, networks=None):
+        with open(path.join(self.base_path, 'go.json'), 'r') as go_f:
+            return json.load(go_f)
+
+class StringDB(object):
+    def __init__(self, host='localhost', port=5432, user='ppin', dbname='stringdb_raw'):
+        self.host = host
+        self.port = port
+        self.user = user
+        self.dbname = dbname
+
+    def connect(self):
+        self.conn = psycopg2.connect(host=self.host, port=self.port, user=self.user, dbname=self.dbname)
+
+    def disconnect(self):
+        if self.conn is not None:
+            self.conn.close()
+            self.conn = None
+
+    def __enter__(self):
+        self.connect()
+        return self
+
+    def __exit__(self):
+        self.disconnect()
+
+    def get_proteins(self, species_id):
+        with self.conn.cursor() as cursor:
+            cursor.execute("""
+                select protein_id, preferred_name
+                from items.proteins
+                where species_id = %s;
+                """,
+                (species_id,))
```

Comment:
Also this modules used to have a params args, so you can send a query and a dict to an execute command, so you won't have problems with types

Reply:
Params args? What problems?

Reply:
To create a query with variables you can write something like
```
sql = """
        SELECT protein_id, preferred_name
        FROM items.proteins
        WHERE species_id = %(specie_id)s
"""
params = {'specie_id': specie_id}
cur.execute(sql, params)
```
You can avoid problems with types and is very useful if one of your variables is a list